ts/src/profiler.ts

```diff
@@ -13,47 +13,358 @@
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    const [body, response] = await this.request(options);
+    if (!hasHttpStatusCode(response)) {
+      throw new Error('Server response missing status information.');
+    }
+    if (isErrorResponseStatusCode(response.statusCode)) {
+      let message: number|string = response.statusCode;
+      // tslint:disable-next-line: no-any
+      if ((response as any).statusMessage) {
+        message = response.statusMessage;
+      }
+      throw new Error(message.toString());
+    }
+    if (!isRequestProfile(body)) {
+      throw new Error(`Profile not valid: ${body}.`);
+    }
+    return body;
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, like a problem collecting or uploading the
+   * profile, an error will be thrown.
+   *
+   * Public to allow for testing.
+   */
+  async profileAndUpload(prof: RequestProfile): Promise<void> {
+    prof = await this.profile(prof);
+    prof.labels = this.profileLabels;
+
+    const options = {
+      method: 'PATCH',
+      uri: API + '/' + prof.name,
+      body: prof,
+      json: true,
+    };
+    const [body, response] = await this.request(options);
+    if (!hasHttpStatusCode(response)) {
+      throw new Error(
+          'Server response missing status information when attempting to upload profile.');
+    }
+    if (isErrorResponseStatusCode(response.statusCode)) {
+      let message: number|string = response.statusCode;
+      // tslint:disable-next-line: no-any
+      if ((response as any).statusMessage) {
+        message = response.statusMessage;
+      }
+      throw new Error(`Could not upload profile: ${message}.`);
+    }
+  }
+
+  /**
+   * Collects a profile of the type specified by profileType field of prof.
+   * If any problem is encountered, for example the profileType is not
+   * recognized or profiling is disabled for the specified profileType, an
+   * error will be thrown.
+   *
+   * Public to allow for testing.
+   */
+  async profile(prof: RequestProfile): Promise<RequestProfile> {
+    switch (prof.profileType) {
+      case ProfileTypes.Wall:
+        return await this.writeTimeProfile(prof);
+      case ProfileTypes.Heap:
+        return this.writeHeapProfile(prof);
+      default:
+        throw new Error(`Unexpected profile type ${prof.profileType}.`);
+    }
+  }
+
+  /**
+   * Collects a time profile, converts profile to compressed, base64 encoded
+   * string, and puts this string in profileBytes field of prof.
+   *
+   * Public to allow for testing.
+   */
+  async writeTimeProfile(prof: RequestProfile): Promise<RequestProfile> {
+    if (!this.timeProfiler) {
+      throw Error('Cannot collect time profile, time profiler not enabled.');
+    }
+    const durationMillis = parseDuration(prof.duration);
```

Comment:
By the way, did you verify that the parse-duration module supports the text duration format used by the Google API JSON responses?

Reply:
Yes -- the responses have formats like '5h5m10s' and that is parsed correctly.
The only possible problem was that this module didn't interpret 'u' as microseconds,  but I didn't see any responses which used 'u'.